subsys/settings/src/settings_its.c

```diff
@@ -0,0 +1,281 @@
+		LOG_ERR("Invalid index %d in ITS metadata", *(int *)back_end);
+		return 0;
+	}
+
+	memcpy(data, entries[*(int *)back_end].value, len);
+
+	/*
+	 * Callback expects return value of the number of bytes read
+	 */
+	return entries[*(int *)back_end].val_len;
+}
+
+static int settings_its_load(struct settings_store *cs, const struct settings_load_arg *arg)
+{
+	int ret;
+
+	for (int i = 0; i < entries_count; i++) {
+
+		if (strnlen(entries[i].name, SETTINGS_MAX_NAME_LEN) != 0) {
+
+			/*
+			 * Pass the key to the settings handler with it's index as an argument,
+			 * to be read during callback function later.
+			 */
+			ret = settings_call_set_handler(entries[i].name, entries[i].val_len,
+							settings_its_read_fn, (void *)&i,
+							(void *)arg);
+			if (ret) {
+				return ret;
+			}
+		}
+	}
+
+	return 0;
+}
+
+static int settings_its_save(struct settings_store *cs, const char *name, const char *value,
+			     size_t val_len)
+{
+	if (entries_count >= CONFIG_SETTINGS_TFM_ITS_NUM_ENTRIES) {
+		LOG_ERR("%s: Max settings reached: %d", __func__,
+			CONFIG_SETTINGS_TFM_ITS_NUM_ENTRIES);
+		return -ENOMEM;
+	}
+
+	if (val_len > SETTINGS_MAX_VAL_LEN) {
+		LOG_ERR("%s: Invalid settings size - val_len: %d", __func__, val_len);
+		return -EINVAL;
+	}
+
+	int index;
+	bool delete;
+
+	/* Find out if we are doing a delete */
+	delete = ((value == NULL) || (val_len == 0));
+
+	/*
+	 * Search metadata to see if entry already exists. Array is compacted, so first blank entry
+	 * signals end of settings.
+	 */
+	for (index = 0; index < CONFIG_SETTINGS_TFM_ITS_NUM_ENTRIES; index++) {
+		if (strncmp(entries[index].name, name, SETTINGS_MAX_NAME_LEN) == 0) {
+			break;
+		} else if (entries[index].val_len == 0) {
+
+			/* Setting already deleted */
+			if (delete) {
+				LOG_DBG("%s: %s Already deleted!", __func__, name);
+				return 0;
+			}
+
+			/* New setting being entered */
+			entries_count++;
+			break;
+		}
+	}
+
+	LOG_DBG("ITS Save - index %d: name %s, val_len %d", index, name, val_len);
+
+	if (delete) {
+		/* Clear metadata */
+		memset(entries[index].name, 0, SETTINGS_MAX_NAME_LEN);
+		memset(entries[index].value, 0, SETTINGS_MAX_VAL_LEN);
+		entries[index].val_len = 0;
+
+		/* If setting not at end of array, shift entries */
+		if (index < entries_count - 1) {
+			memcpy(&entries[index], &entries[index + 1],
+			       (entries_count - index - 1) * sizeof(struct setting_entry));
+			/* Remove duplicate entry at end of array */
+			memset(&entries[entries_count - 1], 0, sizeof(struct setting_entry));
+		}
+
+		entries_count--;
+	} else {
+		/* Update metadata */
+		strncpy(entries[index].name, name, SETTINGS_MAX_NAME_LEN);
+		memcpy(entries[index].value, value, val_len);
+		entries[index].val_len = val_len;
+	}
+
+	k_work_schedule(&worker, K_MSEC(CONFIG_SETTINGS_TFM_ITS_LAZY_PERSIST_DELAY_MS));
+
+	return 0;
+}
+
+void worker_persist_entries_struct_fn(struct k_work *work)
+{
+	k_mutex_lock(&worker_mutex, K_FOREVER);
```

Comment:
I don't think this usage of the mutex is enough. It does not prevent `entries[]` to be modified (calling `settings_its_save()`) while `store_entries()` is executed and that can cause corruption problems.
You can keep mutex lock/unlock in this function, but you also need to use the same mutex in `settings_its_save()`.